### PR TITLE
Switch queued_retry processor to the new internal data structure

### PR DESCRIPTION
**Pull request by dmitryax**

### CODE REVIEW

File: processor/metrics.go [hunk 1]
```diff
@@ -172,10 +172,30 @@ func ServiceNameForResource(resource data.Resource) string {
...
+		} else if serviceName != sn {
+
+			// If there are more then one resources in the TraceData batch "<multiple-services>"
+			// keyword is set. To avoid this, batch processor should be used.
+			serviceName = "<multiple-services>"
+			return
+		}
+	}
+	return
+}
```
Comment (bogdandrutu) [inline processor/metrics.go:191]: I think we should simply report the metric for every service name.

File: processor/queuedprocessor/queued_processor.go [hunk 2]
```diff
@@ -109,27 +113,27 @@ func (sp *queuedSpanProcessor) Stop() {
...
 	if !addedToQueue {
 		// TODO: in principle this may not end in data loss because this can be
 		// in the same call stack as the receiver, ie.: the call from the receiver
 		// to here is synchronous. This means that actually it could be proper to
 		// record this as "refused" instead of "dropped".
-		sp.onItemDropped(item, statsTags)
+		sp.onItemDropped(item, statsTags, spanCount)
 	} else {
-		obsreport.ProcessorTraceDataAccepted(ctx, td)
+		obsreport.ProcessorTraceDataAccepted(ctx, spanCount)
```
Comment (bogdandrutu) [inline processor/queuedprocessor/queued_processor.go:136]: the calls to obsreport can be combined 2 of each. Maybe add a todo?

File: processor/metrics.go (hunk 1 above)
Comment (dmitryax) [reply processor/metrics.go:191]: but it's not going to be accurate. for example, if we have a batch with two services, 2 spans from `service1` and 3 spans from `service2`, we will report:
```
otelcol_batches_received{processor="queued_retry",service="service1"} 5
otelcol_batches_received{processor="queued_retry",service="service2"} 5
```

Comment (bogdandrutu) [reply processor/metrics.go:191]: no, we should report:
otelcol_batches_received{processor="queued_retry",service="service1"} 2
otelcol_batches_received{processor="queued_retry",service="service2"} 3

Comment (dmitryax) [reply processor/metrics.go:191]: right. but it's not possible for `sent` and `dropped` metrics, otherwise we wound need to break down the batch before sending to downstream

Comment (bogdandrutu) [reply processor/metrics.go:191]: I am confused why is not possible. Can you clarify?

Comment (dmitryax) [reply processor/metrics.go:191]: Trying to explain why it's not possible I realized that it's very possible :)

Before sending to downsteam we can save how many spans each service has, then based on the downstream response we emit dropped or sent metrics for all the services

I'll add this change.

Comment (dmitryax) [reply processor/metrics.go:191]: @bogdandrutu I can do what I described for span count metrics ^ But for batches it seems to be actually impossible because one batch (+1 to metric counter) now can contain more than one service. 

In my example I confused batch and span count metrics: if we get a batch with two services, 2 spans from service1 and 3 spans from service2 we can expose metrics:
```
otelcol_spans_received{processor="queued_retry",service="service1"} 2
otelcol_spans_received{processor="queued_retry",service="service2"} 3
```
but cannot expose `otelcol_batches_received` with service tag, so we can either  put `"<multiple-services>"` as the service tag value (in that case i'd do the same for span count metrics as well). Or we can just drop service tag and expose:
```
otelcol_batches_received{processor="queued_retry",source_format="oc_trace"} 
```

What do you think?

**Comment by bogdandrutu** [comment]
https://github.com/open-telemetry/opentelemetry-collector/pull/826 is split from this to allow other components to be changed and not blocked by this PR.

File: processor/metrics.go (hunk 1 above)
Comment (dmitryax) [reply processor/metrics.go]: @bogdandrutu I implemented the approach I proposed above, appeared to be a bit tricky. Please take a look once you have a chance

File: processor/queuedprocessor/queued_processor.go (hunk 2 above)
Comment (dmitryax) [reply processor/queuedprocessor/queued_processor.go]: Sorry, I didn't get it. Which calls can be combined?

**Comment by dmitryax** [comment]
@bogdandrutu PTAL

**Comment by bogdandrutu** [comment]
Please rebase, then I will look.

**Comment by dmitryax** [comment]
@bogdandrutu rebased

File: consumer/pdata/trace.go [hunk 3]
```diff
@@ -78,6 +78,34 @@ func (td Traces) SpanCount() int {
 	return spanCount
 }
 
+// SpanCountByResourceStringAttribute calculates the number of spans by resource specified by
+// provided string attribute attrKey.
+func (td Traces) SpanCountByResourceStringAttribute(attrKey string) map[string]int {
```
Comment (bogdandrutu) [inline consumer/pdata/trace.go:83]: I would not put this in the main API and expose publicly yet. If there are other places where we use this we can decide to move later.

File: internal/collector/telemetry/telemetry.go [hunk 4]
```diff
@@ -31,14 +38,70 @@ const (
...
+	metricsLevelPtr  *string
+	metricsPortPtr   *uint
+	metricsPrefixPtr *string
+
+	useLegacyMetricsPtr *bool
+	useNewMetricsPtr    *bool
+	addInstanceIDPtr    *bool
+)
+
+func Flags(flags *flag.FlagSet) {
```
Comment (bogdandrutu) [inline internal/collector/telemetry/telemetry.go:52]: Not sure why these changes are here.

File: processor/metrics.go [hunk 5]
```diff
@@ -61,17 +62,39 @@ var (
...
+
+func NewSpanCountStats(td pdata.Traces, processorName string) *SpanCountStats {
+	scm := &SpanCountStats{
+		processorName: processorName,
+		allSpansCount: td.SpanCount(),
+	}
+	if serviceTagsEnabled() {
+		scm.serviceSpansCounts = td.SpanCountByResourceStringAttribute(conventions.AttributeServiceName)
+		scm.isDetailed = true
+	}
```
Comment (bogdandrutu) [inline processor/metrics.go:82]: Are you worried that this has a big overhead or why did you try to avoid alway doing this?

Comment (dmitryax) [reply processor/metrics.go:82]: Yes, I'm worried that it's unnecessary overhead and this feature is disabled most of the times I guess. I assume not many users enable `--metric-level=DETAILED`. Not worrying about it makes the implementation cleaner tho.

File: internal/collector/telemetry/telemetry.go (hunk 4 above)
Comment (dmitryax) [reply internal/collector/telemetry/telemetry.go:52]: we need to move flags somewhere from `service` package in order to use `metricsLevel` from  `processor/metrics.go:serviceTagsEnabled `

File: consumer/pdata/trace.go (hunk 3 above)
Comment (dmitryax) [reply consumer/pdata/trace.go:83]: Makes sense

File: processor/metrics.go [hunk 6]
```diff
@@ -61,17 +62,39 @@ var (
...
+	isDetailed         bool
+}
+
+func NewSpanCountStats(td pdata.Traces, processorName string) *SpanCountStats {
+	scm := &SpanCountStats{
+		processorName: processorName,
+		allSpansCount: td.SpanCount(),
+	}
+	if serviceTagsEnabled() {
+		scm.serviceSpansCounts = td.SpanCountByResourceStringAttribute(conventions.AttributeServiceName)
```
Comment (tigrannajaryan) [inline processor/metrics.go:80]: Do we really need this? I suspect this is going to have significant performance impact. I would like to see a perf showing how much time and memory this takes and assess if this is needed at all.
Why do we need to collect per service stats? Is that a collector job? Is this a queue's job?
Looks like we are doing the job that the backend should be doing. If we really need it let's make it a separate processor.
I may be missing something, please clarify.

Comment (tigrannajaryan) [reply processor/metrics.go:80]: Just to be clear. I understand this was an existing functionality. But I do not understand why did we have this in the queued_retry processor.

Comment (dmitryax) [reply processor/metrics.go:80]: Probably because the queued_retry processor is the step where we expect to drop spans. So we want to show how many of them were received and dropped. But I agree that showing which service was affected by drops is not that super-important. And I'd rather simplify the implementation and do not report service based metrics. 

@pjanotti do you know how actively these service grouped metrics being used/

cc @bogdandrutu

Comment (pjanotti) [reply processor/metrics.go:80]: Yes, some have used it in the past to identify which services were generating expected high volume of data. That said this is not available on the default metric configuration one has to select "detailed" to see them.

Comment (tigrannajaryan) [reply processor/metrics.go:80]: As discussed offline we are OK to keep this functionality since it is not enabled by default.

File: consumer/pdata/trace.go (hunk 3 above)
Comment (dmitryax) [reply consumer/pdata/trace.go]: Moved to the package where it's being used and made it internal there

**Comment by dmitryax** [comment]
@bogdandrutu PTAL

File: internal/collector/telemetry/telemetry.go (hunk 4 above)
Comment (bogdandrutu) [reply internal/collector/telemetry/telemetry.go:52]: I think these levels should be part of the host if you want to expose them. Don't want to block the PR, but I think we do spread this metrics levels which is something that may cause us problems in the future.
